### PR TITLE
fix: Preserve empty strings in multipart/form-data requests (#4204)

### LINKED ISSUE #4204: Bug: Empty strings in multipart/form-data requests are converted to None
### Description

Empty strings are parsed correctly when using `application/x-www-form-urlencoded`. However, when using `multipart/form-data`, they instead get converted to `None`.

This is due to `litestar/_multipart.py:112`:

```
if isinstance(data, UploadFile):
    await data.seek(0)
    fields[segment.name].append(data)
elif data:  # here, data is a bytearray
    fields[segment.name].append(data.decode(segment.charset or "utf-8"))
else:
    fields[segment.name].append(None)
```

Because `data` is tested using boolean conversion instead of explicit comparison,`''` (which converts to the empty bytearray) will fail the check and the `else` branch will be taken.

I'm willing to create a PR for this, but it would be a breaking change.

### URL to code causing the issue

_No response_

### MCVE

```python
from typing import Annotated

from litestar import Litestar
from litestar import post
from litestar.enums import RequestEncodingType
from litestar.params import Body
from pydantic import BaseModel


class Data(BaseModel):
    value: str | None


@post("/f", sync_to_thread=False)
def f(data: Annotated[Data, Body(media_type=RequestEncodingType.URL_ENCODED)]) -> bool:
    return data.value is not None


@post("/g", sync_to_thread=False)
def g(data: Annotated[Data, Body(media_type=RequestEncodingType.MULTI_PART)]) -> bool:
    return data.value is not None


app = Litestar(route_handlers=[f, g])
```

### Steps to reproduce

```
>>> import httpx
>>> httpx.post('http://localhost:8000/f', data={'value': ''}).text
'true'
>>> httpx.post('http://localhost:8000/g', data={'value': ''}, files={'file': ''}).text
'false'
```

### Screenshots

_No response_

### Logs

```text

```

### Litestar Version

2.16.0

### Platform

- [ ] Linux
- [x] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

### PR DESCRIPTION
**Pull request by doubledare704**
## Description
- Fix issue #4204 where empty strings in multipart forms were converted to None
- Change boolean check to length check in _multipart.py line 114
- Empty bytearrays now decode to empty strings instead of becoming None
- Add comprehensive tests for empty string handling in multipart forms
- Raw multipart content parsing works correctly
- Existing functionality remains unaffected

### CODE REVIEW

File: tests/unit/test_kwargs/test_multipart_data.py [hunk 1]
```diff
@@ -597,3 +598,93 @@ def test_invalid_multipart_raises_client_error() -> None:
...
+@post("/test-empty-url-encoded", sync_to_thread=False)
+def test_empty_url_encoded(
+    data: Annotated[EmptyStringTestData, Body(media_type=RequestEncodingType.URL_ENCODED)],
+) -> bool:
+    return data.value is not None
+
+
+@post("/test-empty-multipart", sync_to_thread=False)
+def test_empty_multipart(data: Annotated[EmptyStringTestData, Body(media_type=RequestEncodingType.MULTI_PART)]) -> bool:
+    return data.value is not None
```
Comment (provinzkraut) [inline tests/unit/test_kwargs/test_multipart_data.py]: These should be included in the test case

File: litestar/utils/predicates.py [hunk 2]
```diff
@@ -294,3 +295,30 @@ def is_undefined_sentinel(value: Any) -> bool:
         A boolean.
     """
     return any(v is value for v in UNDEFINED_SENTINELS)
+
+
+def is_string_type(annotation: Any) -> bool:
```
Comment (provinzkraut) [inline litestar/utils/predicates.py]: Not necessary as this functionality already exists on `FieldDefinition`, e.g. using `is_subclass_of` / `has_inner_subclass_of`

Comment (doubledare704) [reply litestar/utils/predicates.py]: removed

File: tests/unit/test_kwargs/test_multipart_data.py (hunk 1 above)
Comment (doubledare704) [reply tests/unit/test_kwargs/test_multipart_data.py]: moved to test case

File: tests/unit/test_kwargs/test_multipart_data.py [hunk 3]
```diff
@@ -597,3 +619,93 @@ def test_invalid_multipart_raises_client_error() -> None:
             headers={"Content-Type": "multipart/form-data; charset=utf-8; boundary=20b303e711c4ab8c443184ac833ab00f"},
         )
         assert response.status_code == HTTP_400_BAD_REQUEST
+
+
+# Test for GitHub issue #4204: Empty strings in multipart/form-data should be preserved
+class EmptyStringTestData(BaseModel):
```
Comment (provinzkraut) [inline tests/unit/test_kwargs/test_multipart_data.py]: Why is this testing a Pydantic model? This issue has nothing to do with Pydantic

File: tests/unit/test_kwargs/test_multipart_data.py [hunk 4]
```diff
@@ -504,14 +504,27 @@ class ProductForm:
     optional_with_default: Optional[int] = None
 
 
-def test_multipart_handling_of_none_values() -> None:
+def test_multipart_handling_of_optional_values() -> None:
+    """Test that multipart forms handle optional fields correctly.
```
Comment (provinzkraut) [inline tests/unit/test_kwargs/test_multipart_data.py:509]:
```suggestion
    # https://github.com/litestar-org/litestar/issues/4204
    """Test that multipart forms handle optional fields correctly.
```

File: tests/unit/test_kwargs/test_multipart_data.py [hunk 5]
```diff
@@ -597,3 +618,100 @@ def test_invalid_multipart_raises_client_error() -> None:
...
+
+        # Test multipart form
+        response_multipart = client.post("/consistency-test", data={"value": ""}, files={"dummy": ""})
+        multipart_result = response_multipart.json()
+
+        # Both should return empty string, not None
+        assert url_result["value"] == ""
+        assert multipart_result["value"] == ""
+        assert url_result["value_type"] == "str"
+        assert multipart_result["value_type"] == "str"
```
Comment (provinzkraut) [inline tests/unit/test_kwargs/test_multipart_data.py]: This is already covered by the other tests. You should try to keep these units tests minimal and focused; They ideally should only test *one* thing. What this test wants to show is that the multipart / url encoded form data decodes the same, so there should only be one assert needed for this.

This makes the tests easier to follow, since their intention is more clear.

File: tests/unit/test_kwargs/test_multipart_data.py [hunk 6]
```diff
@@ -597,3 +618,100 @@ def test_invalid_multipart_raises_client_error() -> None:
...
+
+        # Results should be identical
+        assert url_result == multipart_result
+
+
+def test_multipart_empty_string_with_raw_content() -> None:
```
Comment (provinzkraut) [inline tests/unit/test_kwargs/test_multipart_data.py]: This test is unnecessary. It only proves that the manually generated multipart payload is equivalent to the one's generated by httpx. We already assume this is the case anyway, and we're not trying to prove that the parser is working correctly in this test case.

File: tests/unit/test_kwargs/test_multipart_data.py [hunk 7]
```diff
@@ -597,3 +618,100 @@ def test_invalid_multipart_raises_client_error() -> None:
             headers={"Content-Type": "multipart/form-data; charset=utf-8; boundary=20b303e711c4ab8c443184ac833ab00f"},
         )
         assert response.status_code == HTTP_400_BAD_REQUEST
+
+
+# Test for GitHub issue #4204: Empty strings in multipart/form-data should be preserved
```
Comment (provinzkraut) [inline tests/unit/test_kwargs/test_multipart_data.py]:
```suggestion
# https://github.com/litestar-org/litestar/issues/4204
```

File: tests/unit/test_kwargs/test_multipart_data.py [hunk 8]
```diff
@@ -597,3 +618,100 @@ def test_invalid_multipart_raises_client_error() -> None:
...
+def test_empty_strings_preserved_in_multipart_forms() -> None:
+    """Test that empty strings are preserved in multipart forms and not converted to None.
+
+    This test addresses GitHub issue #4204 where empty strings in multipart/form-data
+    requests were being converted to None instead of being preserved as empty strings.
+    The test focuses on core multipart parsing behavior, not serialization library specifics.
```
Comment (provinzkraut) [inline tests/unit/test_kwargs/test_multipart_data.py]:
```suggestion
    This test addresses GitHub issue #4204 where empty strings in multipart/form-data
    requests were being converted to None instead of being preserved as empty strings.
    The test focuses on core multipart parsing behavior, not serialization library specifics.
```
this doesn't really add anything, a short comment and a link to the issue is fine

File: tests/unit/test_kwargs/test_multipart_data.py [hunk 9]
```diff
@@ -597,3 +618,100 @@ def test_invalid_multipart_raises_client_error() -> None:
...
+
+    @post("/test-form")
+    async def form_handler(request: Request) -> dict[str, Any]:
+        """Handler that directly tests form parsing without any serialization library."""
+        data = await request.form()
+        return {
+            "value": data.get("value"),
+            "value_type": type(data.get("value")).__name__,
+            "value_is_none": data.get("value") is None,
+        }
```
Comment (provinzkraut) [inline tests/unit/test_kwargs/test_multipart_data.py]:
```suggestion
        assert data.get("value") == ""
```

File: tests/unit/test_kwargs/test_multipart_data.py [hunk 10]
```diff
@@ -597,3 +618,100 @@ def test_invalid_multipart_raises_client_error() -> None:
...
+        # Both should preserve empty strings, not convert to None
+        assert url_result["value"] == ""
+        assert multipart_result["value"] == ""
+        assert url_result["value_type"] == "str"
+        assert multipart_result["value_type"] == "str"
+        assert url_result["value_is_none"] is False
+        assert multipart_result["value_is_none"] is False
+
+        # Results should be identical (consistency between form types)
+        assert url_result == multipart_result
```
Comment (provinzkraut) [inline tests/unit/test_kwargs/test_multipart_data.py]: You already have an extra test case below that checks if multipart and url-encoded are the same. There is no need to test the same behaviour multiple times in multiple different tests. See my earlier explanation as to why it's better to keep tests small and focused.

**Review by provinzkraut — CHANGES_REQUESTED** [review]
This already looks better. Still, a couple of suggestions for the tests

**Comment by gsakkis** [comment]
After upgrading to 2.18 some of my tests broke due to this change. Ultimately it's an issue on my end but it took some time to figure it out, plus there are some difference between DTO (PydanticDTO specifically) and non-DTO (Pydantic BaseModel). 

In short:
- There is a `MyForm` model containing an UploadFile and a nullable field.
- In some tests I was passing `{field: None}` as form data and apparently http clients (or at least httpx) converts None to empty string.
- In 2.17 this works (i.e. the empty string is validated successfully as None) both when annotating the data as `MyForm` and as `DTOData[MyForm]`.
- In 2.18 it fails with
  - HTTP 400 (`litestar.ValidationException`) if annotated as `MyForm`
  - HTTP 500 (`msgspec.ValidationError`) if annotated as `DTOData[MyForm]` with `dto=PydanticDTO[MyForm]`
- Unrelated to this PR but while playing around I discovered that (both in 2.17 and 2.18) the string `"null"` is surprisingly converted to None in the DTO case (but not in the plain BaseModel case). Not sure if this is a bug, a feature or neither.
   - Update: it's even stranger, apparently it's case insensitive - `"NULL"` and `"Null"` are also converted to None.

I can post a MCVE if interested.

**Comment by euri10** [comment]
> I can post a MCVE if interested.

i think you should

**Comment by gsakkis** [comment]
- Server
```python
from datetime import date
from typing import Annotated, Any

from litestar import Litestar, post
from litestar.datastructures import UploadFile
from litestar.dto import DTOData
from litestar.enums import RequestEncodingType
from litestar.params import Body
from litestar.plugins.pydantic import PydanticDTO
from pydantic import BaseModel


class MyForm(BaseModel, arbitrary_types_allowed=True):
    script: UploadFile
    updated_on: date | None = None


@post("/")
async def process(
    data: Annotated[MyForm, Body(media_type=RequestEncodingType.MULTI_PART)],
) -> dict[str, Any]:
    return {"filename": data.script.filename, "updated_on": data.updated_on}


@post("/dto", dto=PydanticDTO[MyForm])
async def process_dto(
    data: Annotated[DTOData[MyForm], Body(media_type=RequestEncodingType.MULTI_PART)],
) -> dict[str, Any]:
    builtins = data.as_builtins()
    return {
        "filename": builtins["script"].filename,
        "updated_on": builtins["updated_on"],
    }


app = Litestar(route_handlers=[process, process_dto])
```

- Client
```python
import httpx

with open("upload.py", "rb") as f:
    files = {"script": f}

    for data in {}, {"updated_on": None}, {"updated_on": "null"}:
        print(f"\n--- data={data} ---\n")
        resp = httpx.post("http://localhost:8000", files=files, data=data)
        print(resp.text)
        print("-" * 120)
        resp = httpx.post("http://localhost:8000/dto", files=files, data=data)
        print(resp.text)
```

- Output with 2.17
```
--- data={} ---

INFO - 2025-10-08 10:12:00,484 - httpx - _client - HTTP Request: POST http://localhost:8000 "HTTP/1.1 201 Created"
{"filename":"upload.py","updated_on":null}
------------------------------------------------------------------------------------------------------------------------
INFO - 2025-10-08 10:12:00,489 - httpx - _client - HTTP Request: POST http://localhost:8000/dto "HTTP/1.1 201 Created"
{"filename":"upload.py","updated_on":null}

--- data={'updated_on': None} ---

INFO - 2025-10-08 10:12:00,495 - httpx - _client - HTTP Request: POST http://localhost:8000 "HTTP/1.1 201 Created"
{"filename":"upload.py","updated_on":null}
------------------------------------------------------------------------------------------------------------------------
INFO - 2025-10-08 10:12:00,500 - httpx - _client - HTTP Request: POST http://localhost:8000/dto "HTTP/1.1 201 Created"
{"filename":"upload.py","updated_on":null}

--- data={'updated_on': 'null'} ---

INFO - 2025-10-08 10:12:00,509 - httpx - _client - HTTP Request: POST http://localhost:8000 "HTTP/1.1 400 Bad Request"
{"status_code":400,"detail":"Validation failed for POST /","extra":[{"message":"Input should be a valid date or datetime, input is too short","key":"updated_on"}]}
------------------------------------------------------------------------------------------------------------------------
INFO - 2025-10-08 10:12:00,515 - httpx - _client - HTTP Request: POST http://localhost:8000/dto "HTTP/1.1 201 Created"
{"filename":"upload.py","updated_on":null}
```

- Output with 2.18
```

--- data={} ---

INFO - 2025-10-08 10:11:23,666 - httpx - _client - HTTP Request: POST http://localhost:8000 "HTTP/1.1 201 Created"
{"filename":"upload.py","updated_on":null}
------------------------------------------------------------------------------------------------------------------------
INFO - 2025-10-08 10:11:23,674 - httpx - _client - HTTP Request: POST http://localhost:8000/dto "HTTP/1.1 201 Created"
{"filename":"upload.py","updated_on":null}

--- data={'updated_on': None} ---

INFO - 2025-10-08 10:11:23,681 - httpx - _client - HTTP Request: POST http://localhost:8000 "HTTP/1.1 400 Bad Request"
{"status_code":400,"detail":"Validation failed for POST /","extra":[{"message":"Input should be a valid date or datetime, input is too short","key":"updated_on"}]}
------------------------------------------------------------------------------------------------------------------------
INFO - 2025-10-08 10:11:23,688 - httpx - _client - HTTP Request: POST http://localhost:8000/dto "HTTP/1.1 500 Internal Server Error"
{"status_code":500,"detail":"Internal Server Error"}

--- data={'updated_on': 'null'} ---

INFO - 2025-10-08 10:11:23,694 - httpx - _client - HTTP Request: POST http://localhost:8000 "HTTP/1.1 400 Bad Request"
{"status_code":400,"detail":"Validation failed for POST /","extra":[{"message":"Input should be a valid date or datetime, input is too short","key":"updated_on"}]}
------------------------------------------------------------------------------------------------------------------------
INFO - 2025-10-08 10:11:23,699 - httpx - _client - HTTP Request: POST http://localhost:8000/dto "HTTP/1.1 201 Created"
{"filename":"upload.py","updated_on":null}
```
